src/PIL/HeifImagePlugin.py

```diff
@@ -0,0 +1,36 @@
+import io
+
+from . import Image, ImageFile
+
+try:
+    import pyheif
+except ImportError:
+    pyheif = None
+
+def _accept(prefix):
+    check = pyheif.check(prefix)
+    return check != pyheif.heif_filetype_no
+
+
+class HeifImageFile(ImageFile.ImageFile):
+    format = "HEIF"
+    format_description = "HEIF Image File Format"
+
+    def _open(self):
+        b = self.fp.read()
+        if not _accept(b):
```

Comment:
In general we should not be reading the whole file in _open. It does appear that pyheif does that, which is somewhat disappointing. I'm not sure if it's possible to get around that, except for the initial magic check.